# HG changeset patch
# User Rik <rik@octave.org>
# Date 1449093937 28800
#      Wed Dec 02 14:05:37 2015 -0800
# Branch stable
# Node ID a8ee668e7fd7cae9b821bc9b8676df46ff29db13
# Parent  51ab9145f6e677e46e04765ef0f3df894f6277f7
Fix kbhit and pause on Windows systems.

* configure.ac: Add check for function _getch.

* sysdep.cc (octave_kbhit): Use _getch, _kbhit on Windows platforms to
correctly implement kbhit functionality.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -2277,17 +2277,17 @@ AC_CHECK_FUNCS([getegid geteuid getgid g
 AC_CHECK_FUNCS([getpgrp getpid getppid getpwent getpwuid getuid])
 AC_CHECK_FUNCS([isascii kill])
 AC_CHECK_FUNCS([lgamma lgammaf lgamma_r lgammaf_r])
 AC_CHECK_FUNCS([log1p log1pf pipe])
 AC_CHECK_FUNCS([realpath resolvepath roundl])
 AC_CHECK_FUNCS([select setgrent setpwent siglongjmp strsignal])
 AC_CHECK_FUNCS([tcgetattr tcsetattr tgammaf toascii])
 AC_CHECK_FUNCS([umask waitpid])
-AC_CHECK_FUNCS([_kbhit])
+AC_CHECK_FUNCS([_getch _kbhit])
 
 dnl There are no workarounds in the code for missing these functions.
 AC_CHECK_FUNCS([modf pow sqrt sqrtf], [],
                [AC_MSG_ERROR([Missing function required to build Octave])])
 
 ## exp2, round, tgamma function checks
 AC_LANG_PUSH(C++)
 AC_CHECK_DECLS([exp2, round, tgamma], [], [], [[#include <cmath>]])
diff --git a/libinterp/corefcn/sysdep.cc b/libinterp/corefcn/sysdep.cc
--- a/libinterp/corefcn/sysdep.cc
+++ b/libinterp/corefcn/sysdep.cc
@@ -508,18 +508,25 @@ octave_pclose (FILE *f)
 #endif
 }
 
 // Read one character from the terminal.
 
 int
 octave_kbhit (bool wait)
 {
-#ifdef HAVE__KBHIT
-  int c = (! wait && ! _kbhit ()) ? 0 : std::cin.get ();
+#ifdef HAVE__KBHIT && HAVE__GETCH
+  // This essentially means we are on a Windows system.
+  int c;
+
+  if (wait)
+    c = _getch ();
+  else
+    c = (! _kbhit ()) ? 0 : _getch ();
+
 #else
   raw_mode (true, wait);
 
   // Get current handler.
   octave_interrupt_handler saved_interrupt_handler
     = octave_ignore_interrupts ();
 
   // Restore it, disabling system call restarts (if possible) so the
